### PR TITLE
feat(redis & redis-4): Add db.redis.database_index argument to instrumentation

### PR DESCRIPTION
**Pull request by luke6Lh43**
## Which problem is this PR solving?

In the current version, the **redis** and **redis-4** instrumentation does not send **db.redis.database_index** to OpenTelemetry collector. This was discovered during testing on some application. It may be a problem in a scenario when multiple Redis databases are used and we want to differentiate it in some backend (e.g. on a flow map). 

According to [this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md) document, we should send db.redis.database_index argument to OTel collector/backend.

This PR is related to the issue [#1288](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1288
)

## Short description of the changes

- I have added **db.redis.database_index** instrumentation to both **redis** and **redis-4** packages. 
- Also, it is defaulted to **0** when database index is not specified in Redis client definition.

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-redis/src/internal-types.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@ export interface RedisPluginClientTypes {
   options?: {
     host: string;
     port: string;
+    db: string;
```
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-redis/src/internal-types.ts:20]: According to `@types/redis`, this option can also be `string` which we need to convert to `number` when recording into an attribute:

```js
    db?: string | number | undefined;
```

**Review by blumamir — COMMENTED** [review]
Thanks for submitting this PR.

Please add unittests for these changes.

Also, that database index can be set via client options as you implemented, but it can also be set with the [`select` command](https://redis.io/commands/select/), so if I understand the mechanism correctly, the following code 

```js
const client = redis.createClient();
client.select(1);
client.set('foo', 'bar');
```

Will record `db.redis.database_index` as `0` for `set` command, where the actual database index the command was sent to is `1`.

Instrumentation will probably need to track the active current database index for each client when the user calls `select` and use it to record into the span attribute.

**Comment by blumamir** [comment]
@luke6Lh43 are you still working on this?

**Comment by luke6Lh43** [comment]
> @luke6Lh43 are you still working on this?

Yes, had a bit tough end of 2022 but I'm getting back to work on it.

**Comment by blumamir** [comment]
Hi @luke6Lh43 , any update on this? should we keep the PR open?

File: plugins/node/opentelemetry-instrumentation-redis-4/src/utils.ts [hunk 2]
```diff
@@ -19,10 +19,21 @@ import {
 } from '@opentelemetry/semantic-conventions';
 
 export function getClientAttributes(options: any) {
+
+  let db_redis_database_index;
```
[the remark this answers is not in the document]
Comment (luke6Lh43) [reply plugins/node/opentelemetry-instrumentation-redis-4/src/utils.ts:23]: got it, will adjust it to camelCase convention

File: plugins/node/opentelemetry-instrumentation-redis-4/src/utils.ts [hunk 3]
```diff
@@ -19,10 +19,21 @@ import {
...
+
+      //checking if database index is specified in Redis client (if not, default db_redis_database_index = 0)
+  if (options?.database){
+    db_redis_database_index = options?.database
+  }else{
+    db_redis_database_index = '0'
```
[the remark this answers is not in the document]
Comment (luke6Lh43) [reply plugins/node/opentelemetry-instrumentation-redis-4/src/utils.ts:29]: Correct, I missed that. Will adjust it to int type!

File: plugins/node/opentelemetry-instrumentation-redis-4/src/utils.ts [hunk 4]
```diff
@@ -19,10 +19,21 @@ import {
...
+  }
```
[the remark this answers is not in the document]
Comment (luke6Lh43) [reply plugins/node/opentelemetry-instrumentation-redis-4/src/utils.ts:30]: ok, will test it and adjust if working as expected

**Pull request closed without merging**